### PR TITLE
rachael water class

### PR DESCRIPTION
**Pull request by RachaelGomez**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I created a custom method to choose a spotlight item in the work.rb file
Describe how you approached testing that model method. What edge cases did you come up with? | I tested to make sure that an item would show up and that it was a valid example of work
What are `session` and `flash`? What is the difference between them? | session is the current session in a browser when a person signs in. It lasts until the browser is closed. A flash is a quick message to the user that lasts only as long a get/receive cycle.
What was one thing that you gained more clarity on through this assignment? | relationships and testing
What is the Heroku URL of your deployed application? | https://rachaels-media-ranker-app.herokuapp.com/works

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: test/models/work_test.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  end
+  describe 'spotlight' do
+    it 'spotlight work is instance of work' do
+      @work = works(:movie)
+
+      @works = Work.all
+
+      result = @works.spotlight
+
+      expect(result).must_be_instance_of Work
```
Comment (CheezItMan) [inline test/models/work_test.rb:34]: Just a note, if the variable is only used inside the `it` block.  You don't need to make it an instance variable

```suggestion
      work = works(:movie)

      works = Work.all

      result = Work.spotlight

      expect(result).must_be_instance_of Work
```

File: test/models/work_test.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    end
+  end
+  describe 'sort_votes' do
+    it 'sorts works by votes' do
+      @work = works(:book)
+      @works = Work.all
+
+      result = @works.sort_votes
+
+      expect(result.first.title).must_be_same_as @work.title
```
Comment (CheezItMan) [inline test/models/work_test.rb:75]: `must_be_same_as` compares to see if they are the same object (object_id) and not the same value

```suggestion
      expect(result.first.title).must_equal @work.title
```

File: test/models/work_test.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      expect(result).must_be_instance_of Work
+    end
+    it 'chooses the most voted for piece of work' do
+      @work = works(:book)
+
+      @works = Work.all
+
+      result = @works.spotlight
+
+      expect(result.title).must_be_same_as "Normal People"
```
Comment (CheezItMan) [inline test/models/work_test.rb:43]: must_be_same_as compares to see if they are the same object (object_id) and not the same value

```suggestion
      expect(result.title).must_equal "Normal People"
```

File: test/models/vote_test.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      expect(result).must_equal false
+    end
+  end
+  describe 'relationships' do
+    it 'belongs to a work' do
+    work = works(:movie)
+    user = users(:test_user)
+    @vote = Vote.new(user.id, work.id)
+
+    expect(@vote.work_id).must_be_same_as work.id
```
Comment (CheezItMan) [inline test/models/vote_test.rb:33]: `must_be_same_as` compares to see if they are the same object_id, not the same value.  

Better to do:

```suggestion
    expect(@vote.work_id).must_equal work.id
```

File: test/models/vote_test.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    end
+    it 'belongs to a user' do
+      work = works(:movie)
+      user = users(:test_user)
+      @vote = Vote.new(user.id, work.id)
+
+      expect(vote.user_id).must_be_same_as user.id
```
Comment (CheezItMan) [inline test/models/vote_test.rb:40]: must_be_same_as compares to see if they are the same object_id, not the same value.

```suggestion
      expect(vote.user_id).must_equal user.id
```

File: test/models/vote_test.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      result = @vote.valid?
+
+      expect(result).must_equal false
+    end
+  end
+  describe 'relationships' do
+    it 'belongs to a work' do
+    work = works(:movie)
+    user = users(:test_user)
+    @vote = Vote.new(user.id, work.id)
```
Comment (CheezItMan) [inline test/models/vote_test.rb:31]:
```suggestion
    @vote = Vote.new(user_id: user.id, work_id: work.id)
```

File: test/models/vote_test.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    expect(@vote.work_id).must_be_same_as work.id
+    end
+    it 'belongs to a user' do
+      work = works(:movie)
+      user = users(:test_user)
+      @vote = Vote.new(user.id, work.id)
```
Comment (CheezItMan) [inline test/models/vote_test.rb:38]:
```suggestion
      vote = Vote.new(user.id, work.id)
```

File: test/models/vote_test.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      # Arrange
+      @vote = votes(:vote_movie)
+    end
+
+    it 'is valid when all fields are present' do
+      # Act
+      result = @vote.valid?
+
+      # Assert
+      expect(result).must_equal true
```
Comment (CheezItMan) [inline test/models/vote_test.rb:15]: This is failing because your vote fixture doesn't have a user or work.

See my comment on your fixtures.

File: test/fixtures/votes.yml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+vote_movie:
+  user_id: test_user
+  work_id: movie
```
Comment (CheezItMan) [inline test/fixtures/votes.yml:3]: You don't specify the id just note this this refers to a user or work.

```suggestion
  user: test_user
  work: movie
```

File: test/fixtures/votes.yml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+vote_album:
+  user_id: test_user
+  work_id: album
```
Comment (CheezItMan) [inline test/fixtures/votes.yml:7]:
```suggestion
  user: test_user
  work: album
```

File: test/fixtures/votes.yml [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+vote_book:
+  user_id: test_user
+  work_id: book
```
Comment (CheezItMan) [inline test/fixtures/votes.yml:11]:
```suggestion
  user: test_user
  work: book
```

File: test/fixtures/votes.yml [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+more_votes:
+  user_id: test_user_two
+  work_id: book
```
Comment (CheezItMan) [inline test/fixtures/votes.yml:15]:
```suggestion
  user: test_user_two
  work: book
```

File: test/models/user_test.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
+require "test_helper"
+
+describe User do
```
Comment (CheezItMan) [inline test/models/user_test.rb:3]: Noting no tests on relationships.

File: test/models/homepage_test.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe Homepage do
+  # it "does a thing" do
+  #   value(1+1).must_equal 2
+  # end
+end
```
Comment (CheezItMan) [inline test/models/homepage_test.rb:7]: No tests for homepage?


```suggestion
describe Homepage do
  it "loads properly" do
    get root_path
    
    must_respond_with :success
  end
  
  it "loads with no works" do
    Work.destroy_all
    
    get root_path
    
    must_respond_with :success
  end
end
```

File: test/models/vote_test.rb [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    end
+
+    it 'is not valid when fields are not present' do
+      @vote = Vote.new()
+
+
+      result = @vote.valid?
+
+      expect(result).must_equal false
```
Comment (CheezItMan) [inline test/models/vote_test.rb:24]: You  should also include checks for the validation errors so the proper ones are listed.  Both the fields that are invalid and the error messages.

File: test/models/work_test.rb [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      # Assert
+      expect(result).must_equal true
+    end
+
+    it 'is not valid when fields are not present' do
+      @work = Work.new()
+
+      result = @work.valid?
+
+      expect(result).must_equal false
```
Comment (CheezItMan) [inline test/models/work_test.rb:23]: You should also include tests for which fields are invalid and the error messages.

File: test/models/work_test.rb (hunk 1 above)
Comment (CheezItMan) [inline test/models/work_test.rb:34]: You should also include a check for which works are returned as a result of `top_ten` methods and  spotlight.

File: test/controllers/homepages_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe HomepagesController do
+  # it "does a thing" do
+  #   value(1+1).must_equal 2
+  # end
+end
```
Comment (CheezItMan) [inline test/controllers/homepages_controller_test.rb:7]: Just noting no controller tests.

File: app/views/homepages/index.html.erb [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+          <%= @works.spotlight.title %>
+            <% end %>
+        </h3>
+    </section>
+    <div>
+      <h4> Top 10 Books! </h4>
+        <% if @works.empty? %>
+          <p> The top 10 books will go here! </p>
+      <% elsif %>
+        <% @works.top_ten_books.empty? %>
```
Comment (CheezItMan) [inline app/views/homepages/index.html.erb:17]: Is this supposed to be an elsif?

```suggestion
      <% elsif @works.top_ten_books.empty?  %>
```

File: app/views/homepages/index.html.erb [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+      <h4> Top 10 Movies! </h4>
+        <% if @works.empty? %>
+          <p> The top 10 movies will go here! </p>
+        <% elsif %>
+      <% @works.top_ten_movies.empty? %>
+        <p> Vote on some movies to list the best here! </p>
+        <% else %>
+        <% @works.top_ten_movies.each do |work|  %>
+          <% if work.category == "movie" %>
```
Comment (CheezItMan) [inline app/views/homepages/index.html.erb:35]: Why is this here if the `top_ten_movies` method returns only movies?

File: app/models/work.rb [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def self.top_ten_movies
+    work = Work.all
+    movies = []
+    work.each do |work|
+      if work.category == "movies"
+        movies << work
+      end
+    end
```
Comment (CheezItMan) [inline app/models/work.rb:53]: It's better to use `where`

```suggestion
    movies = Work.where(category: "movie")
```

Also note that you're filtering by the category "movies" instead of "movie".

File: app/models/work.rb [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    return sorted_work
+  end
+
+  def self.spotlight
+    work = Work.all
+    spotlight = work.sort_votes.first
+    return spotlight
+  end
+
+  def self.top_ten_books
```
Comment (CheezItMan) [inline app/models/work.rb:22]: These methods are ok, but you can dry them up.

```ruby
  def self.top_ten(cateogry)
    return Work.where(category: category).sort_by { |work| work.votes.count }.reverse.first(10)
  end
```

Or this one which uses Postgres to sort (google search to get it).  

```ruby
  def self.top_ten(cateogry)
    return Work.where(category: category).order(Arel.sql('COUNT(votes.id)DESC'))
  end
```

File: app/models/work.rb [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def self.top_ten_albums
+    work = Work.all
+    albums = []
+    work.each do |work|
+      if work.category == "albums"
+        albums << work
+      end
+    end
```
Comment (CheezItMan) [inline app/models/work.rb:41]:
```suggestion
    albums = Work.where(category: "album")
```

File: app/models/vote.rb [hunk 23]
```diff
@@ -0,0 +1,7 @@
+class Vote < ApplicationRecord
+  belongs_to :work
+  belongs_to :user
+
+  validates_uniqueness_of :user_id, scope: :work_id
```
Comment (CheezItMan) [inline app/models/vote.rb:5]: 👍

File: app/views/works/index.html.erb [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    <ul> <% @works.sort_votes.each do |work|  %>
+        <% if work.category == "book" %>
+        <li> <%= link_to work.title, work_path(work) %> |
+          <%= link_to "Edit", edit_work_path(work[:id]), method: :get%> |
+          <%= link_to "Delete", work_path(work[:id]), method: :delete, data: {confirm: "Are you sure?"} %>
+        <%= link_to "Upvote", vote_work_path(work[:id]), method: :post %></li>
+          <li> Votes: <%= work.votes.count %></li>
+          <% end %>
+      <% end %>
+    </ul>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:16]: Just a note, this could be worked into a partial since you're doing the same thing 3 times here.

File: app/controllers/works_controller.rb [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+
+    if @work.nil?
+      head :not_found
+      flash.now[:error] = "Something happened. Media not updated."
+      return
+    elsif @work.update(work_params)
+      flash[:success] = "#{@work.title} was successfully updated!"
+      redirect_to works_path
+      render :edit
+      return
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:64]: Missing an else

```suggestion
    elsif @work.update(work_params)
      flash[:success] = "#{@work.title} was successfully updated!"
      redirect_to works_path
      return
    else
      flash[:errors] = @work.errors.messages # put in the flash notices
      render :edit, status: :bad_request
      return
```

File: app/views/layouts/application.html.erb [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    </ul>
+  </nav>
+</header>
+
+<section class="flash-msg__section">
+  <% flash.each do |name, message| %>
+    <strong class="<%= name %>">
+      <%= message %>
+    </strong>
+  <% end %>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:38]: This is just a suggestion to iterate through all the validation errors.

```suggestion
    <% if name == :errors %>
      <%= flash[name].each do |field, messages| %>
        <%= "#{field}: #{messages} %>
      <% end %>
    <% else %>
    <strong class="<%= name %>">
      <%= message %>
    </strong>
   <% end %>
  <% end %>
```

File: app/controllers/works_controller.rb [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+  end
+
+  def show
+    work_id = params[:id]
+    @work = Work.find_by(id: work_id)
+    if @work.nil?
+      redirect_to works_path
+    end
+    if @work.nil?
+      head :not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:13]: Maybe better from now on to redirect and provide a flash notice.

File: config/routes.rb [hunk 28]
```diff
@@ -0,0 +1,26 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
+
+  root 'homepages#index'
+
+  get '/homepages', to: 'homepages#index', as: 'homepages'
```
Comment (CheezItMan) [inline config/routes.rb:7]: You don't need this

```suggestion
```

File: config/routes.rb [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+
+  # get '/works', to: 'works#index', as: 'works'
+  # get '/works/new', to: 'works#new', as: 'new_work'
+  # post '/works', to: 'works#create'
+  # get '/works/:id/edit', to: 'works#edit', as: 'edit_work'
+  # patch '/works/:id', to: 'works#update'
+  # get '/works/:id', to: 'works#show', as: 'work'
+  # delete '/works/:id', to: 'works#destroy'
+  post 'works/:id', to: 'works#vote', as: 'vote_work'
```
Comment (CheezItMan) [inline config/routes.rb:16]: I'd suggest instead making this a nested route.

```ruby
resources :works do 
  resources :votes, only: [:create]
end

File: app/views/homepages/index.html.erb [hunk 30]
```diff
@@ -0,0 +1,57 @@
+<body>
```
Comment (CheezItMan) [inline app/views/homepages/index.html.erb:1]: Body should be in the application.html.erb

**Review by CheezItMan — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ⚠️, sort of, you're not really listing the 3 categories you've got a bug with your top_ten methods.  See my inline comments. 
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️, However I don't have an `edit` link.
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ⚠️ Does not provide meaningful validation errors
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ⚠️ Didn't quite get the same styling with the 3 columns.  Otherwise not bad
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️, for the most part yes.
| 7. The homepage view, all media view, and new works view use partials when appropriate | ⚠️ No partials
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ⚠️, see my suggestions for improved methods.  Yours are kinda broken
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ⚠️ No controller filter
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work |  ⚠️ No controller filter
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️, for the most part, see my inline comments

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️, it's there with one bug
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ⚠️ No relationship tests
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️, see my inline comments for errrors
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️, see my inline comments for missing items
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️, see my inline comments for errors on tests

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |  💛





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Logical/Organized | ✅

## Summary

It's clear you struggled a bit with testing and displaying validation errors and styling to match the site.  You **do** have the basic functionality in place.  Thus this is a yellow.  Take a look at my inline comments and let me know what questions you have.  Specifically take a look at your model methods, testing comments and your flash notices.  These are the things to work on.
